README.md
# Logr

* [Logr](#logr)
  * [What is it? (TL;DR)](#what-is-it-tldr)
  * [Build](#build)
     * [Prerequisetes](#prerequisetes)
     * [Build on linux with Conan](#build-on-linux-with-conan)
     * [Build on windows with Conan](#build-on-windows-with-conan)
  * [Using logr](#using-logr)
  * [Naming](#naming)
  * [License](#license)

## What is it? (TL;DR)

**Project status**: beta-draft.

This is a experimentsl project showing my approach to build a logger interface used in application development. It is a revised version of what I've been using for the last couple of years in various projects.

This is not a self-contained logging library. Consider it as a "frontend" or a "wrapper" for whatever logging routine of your choice (which can be integrated as a backend). It can be used "as is" in your application or it can be adopted for your case or maybe you can just pick some ideas and came up with a better solution.

Consider the following reasoning behing logr design [here](./design/on-logr-design.md).

## Build

By itself logr is a header-only library. So talking about build we mean build tests, examples and benchmarks.

### Prerequisetes

* C++17 compiler.
* CMake.
* Package manager ([conan](https://conan.io/), or [vcpkg](https://github.com/microsoft/vcpkg)).
* Clone repository with submodules (don't forget `--recurse-submodules` option).

Logr relies on "backend" logging library. And it comes with ready to go backends for a couple of libraries which are considered to be provided by package manager.

### Build on linux with Conan

One of the script in `build-scripts` folder can be used.

Build with gcc 9.3 and run tests.

```
./build-scripts/gcc93_conan.sh --run-ctest
```

Or build it manually:
```bash
mkdir build && cd build

# Install conan dependencies.
conan install \
    -s compiler=gcc \
    -s compiler.version=9.3 \
    -s compiler.libcxx=libstdc++11 \
    -s build_type=Release \
    --build missing \
    ..

# Assume you deafult compiler is GCC 9.3.
# Otherwise set it explicitly:
#     export CC=gcc-9
#     export CPP=g++-9
#     export AR=gcc-ar-9
#     export CXX=g++-9

cmake -DCMAKE_BUILD_TYPE=Release ..
cmake --build . -j $(nproc)
ctest -T test
```

### Build on linux with Conan dockers

Using one of the dockers from [docker hub conanio](https://hub.docker.com/u/conanio):

```bash
docker run -ti --user=$(id -u):$(id -g) \
           -v $(pwd):/sources \
           -w /sources \
           -e CONAN_USER_HOME=/sources/_conan \
           conanio/gcc10 \
           ./build-scripts/gcc_default.sh \
               --run-ctest
```

### Build on windows with Conan

Build with msvc16 (aka vs2019, aka vc142):

```
rd /s /q _msvc16_conan
mkdir _msvc16_conan
cd _msvc16_conan

REM Precompiled log4cplus in conan-center seems to be broken
REM so we force its build `-b log4cplus`.
conan install -s compiler="Visual Studio" ^
              -s compiler.version=16 ^
              -s build_type=Release ^
              -s compiler.runtime=MD ^
              --build missing ^
              -b log4cplus ^
              ..

cmake -G "Visual Studio 16 2019" ^
      -DCMAKE_BUILD_TYPE=Release ^
      ..

cmake --build . -j 4 --config Release
ctest -T test
```
## Build with Vcpkg

To build tests, examples and benchmarks the following packages should be installed: glog, spdlog, log4cplus, gtest, benchmark.

Having all dependencies in place one can build the project using the following commands:

```bash
mkdir _vcpkg_build
cd _vcpkg_build

cmake -DCMAKE_BUILD_TYPE=Release -DLOGR_PKG_PROVIDER=vcpkg  -DCMAKE_TOOLCHAIN_FILE=PATH_TO/vcpkg/scripts/buildsystems/vcpkg.cmake ..

cmake --build . -j 4 --config Release
ctest -T test
```

## Using logr

A sample code using spdlogger with logr:

```C++
#include <spdlog/sinks/stdout_sinks.h>

#include <logr/logr.hpp>
#include <logr/spdlog_backend.hpp>

int main()
{
    int x=42, y=0xFFFF;

    logr::spdlog_logger_t<> logger{
        "console",
        std::make_shared< spdlog::sinks::stdout_sink_st >(),
        logr::log_message_level::trace
    };

    logger.info( "Hello World! [raw message]" );

    logger.info( []() {
        return "Hello World! [message-builder by return]";
    } );

    logger.info( []( auto & out ) {
        fmt::format_to( "Hello World! [message-builder by write-to]" );
    } );

    logger.info( [&]( auto & out ) {
        fmt::format_to( out, "X is {}! and Y is {}", x, y );
    } );

    logger.info( [&]( auto & out ) {
        fmt::format_to( out, "Y is {}! and X is {}", y, x );
    } );

    logger.info( [&x]( auto & out ) {
        fmt::format_to( out, "X is {}! and Y is skipped", x );
    } );

    logger.info( [&y]( auto & out ) {
        fmt::format_to( out, "Y is {}! and X is skipped", y );
    } );
}
```

See all [examples](./examples).

## Naming

The original idea behind the library came from [restinio](https://github.com/Stiffstream/restinio) in which a custom logger can be integrated to RESTinio to track its internal events. That is pretty much the major idea around Logr, so `logr = LOGger from Restinio` to show gratitude to great library. And on the other hand it just 4 letters long which is good for namespaces.

## License

Logr is distributed under the BSD 3-Clause [License](./LICENSE).

Libraries used as logger backends are distributed under their license:

* [spdlog](https://github.com/gabime/spdlog).
* [glog](https://github.com/google/glog).
* [log4cplus](https://github.com/log4cplus/log4cplus).


CHANGELOG.md
# Changelog

## v0.2.0 (beta-draft)

* Switch to `spdlog/1.8.2` version in conanfile.
* Make a core logr class independent from traits class.
This allows having the same base class (as a root) for logers based on traits.
* Make adjustments in cmake-scripts for packaging in [conan-center-index](https://github.com/conan-io/conan-center-index).
* Rethinking write-to msg builder wrapped output.
Introduce a wrapper for output in write-to message builders.
This allows a message builder to receive `out` by reference and also by value.
Thus it is a breaking change as previously using `fmt::format_to(out, ...)` doesn't work anymore.
The following ways to use write-to message builders are possible:

```C++
    int x = make_int();
    std::string s = make_string();

    // All the logging is equivalent.

    // Out by value.
    logger.info( [&]( auto out ) {
        format_to( out, "Write-to message builder 1,  int:'{}', str: '{}'", x, s );
    } );
    logger.info( [&]( auto out ) {
        out.format_to( "Write-to message builder 2,  int:'{}', str: '{}'", x, s );
    } );
    logger.info( [&]( auto out ) {
        ::fmt::format_to( out.buf() "Write-to message builder 3,  int:'{}', str: '{}'", x, s );
    } );

    // Out by reference.
    logger.info( [&]( auto & out ) {
        format_to( out, "Write-to message builder 1,  int:'{}', str: '{}'", x, s );
    } );
    logger.info( [&]( auto & out ) {
        out.format_to( "Write-to message builder 2,  int:'{}', str: '{}'", x, s );
    } );
    logger.info( [&]( auto & out ) {
        ::fmt::format_to( out.buf() "Write-to message builder 3,  int:'{}', str: '{}'", x, s );
    } );
```

## v0.1.0 (alpha-draft)

First release (alpha). includes the following

* Three backends: spdlog, glog and log4cplus.
* A set of tests, examples and a number of simple basic benchmarks.


conanfile.py
from conans import ConanFile, CMake, tools
from conans.errors import ConanInvalidConfiguration

class LogrConan(ConanFile):
    name = "logr"
    version = "0.2.0"
    license = "BSD 3-Clause License"
    author = "Nicolai Grodzitski <utromvecherom@gmail.com>"
    url = "https://github.com/ngrodzitski/logr"
    description = "Logger frontend substitution for spdlog, glog, etc for server/desktop applications"
    topics = ("logger", "development", "util", "utils")

    generators = "cmake_find_package", "cmake"
    settings = "os", "compiler", "build_type", "arch"

    build_policy = "missing"

    # If CI wants test then consider building of examples and benchmarks
    # is a kind of test.
    logr_build_test_and_others = tools.get_env("CONAN_RUN_TESTS", False)

    exports_sources = "logr/*", "CMakeLists.txt", "cmake-scripts/*", "LICENSE.txt"

    options = { 'spdlog_backend' : [True, False],
                'glog_backend' : [True, False],
                'log4cplus_backend' : [True, False] }

    default_options = { 'spdlog_backend': True,
                        'glog_backend': True,
                        'log4cplus_backend': True }

    _cmake = None

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        if self.logr_build_test_and_others:
            self.exports_sources += "tests/*", "examples/*", "benchmarks/*"

    def requirements(self):
        self.requires( "fmt/7.1.2" )

    def build_requirements(self):
        if self.options.spdlog_backend:
            self.build_requires( "spdlog/1.8.2" )
            self.build_requires_options["spdlog"].header_only = True

            # For benchmarks in similar conditions,
            # spdlog should do no exception catching.
            self.build_requires_options["spdlog"].no_exceptions = True

        if self.options.glog_backend:
            self.build_requires( "glog/0.4.0" )

        if self.options.log4cplus_backend:
            self.build_requires( "log4cplus/2.0.5" )
            self.build_requires_options["log4cplus"].unicode = False

        self.build_requires( "gtest/1.10.0" )
        self.build_requires( "benchmark/1.5.2" )

    def configure(self):
        minimal_cpp_standard = "17"
        if self.settings.compiler.cppstd:
            tools.check_min_cppstd(self, minimal_cpp_standard)
        minimal_version = {
            "gcc": "7",
            "clang": "7",
            "apple-clang": "10",
            "Visual Studio": "16"
        }
        compiler = str(self.settings.compiler)
        if compiler not in minimal_version:
            self.output.warn(
                "%s recipe lacks information about the %s compiler standard version support" % (self.name, compiler))
            self.output.warn(
                "%s requires a compiler that supports at least C++%s" % (self.name, minimal_cpp_standard))
            return

        version = tools.Version(self.settings.compiler.version)
        if version < minimal_version[compiler]:
            raise ConanInvalidConfiguration("%s requires a compiler that supports at least C++%s" % (self.name, minimal_cpp_standard))

    def _configure_cmake(self):
        if self._cmake:
            return self._cmake

        self._cmake = CMake(self)
        self._cmake.definitions['LOGR_WITH_SPDLOG_BACKEND'] = self.options.spdlog_backend
        self._cmake.definitions['LOGR_WITH_GLOG_BACKEND'] = self.options.glog_backend
        self._cmake.definitions['LOGR_WITH_LOG4CPLUS_BACKEND'] = self.options.log4cplus_backend

        self._cmake.definitions['LOGR_BUILD_TESTS'] = self.logr_build_test_and_others
        self._cmake.definitions['LOGR_BUILD_EXAMPLES'] = self.logr_build_test_and_others
        self._cmake.definitions['LOGR_BUILD_BENCHMARKS'] = self.logr_build_test_and_others

        if self.settings.compiler == "gcc" or self.settings.compiler == "clang":
            self._cmake.definitions['EXPLICIT_LIBCXX'] = self.settings.compiler.libcxx

        if self.settings.compiler == "Visual Studio":
            rt = self.settings.compiler.runtime
            self._cmake.definitions['EXPLICIT_STATIC_RUNTIME'] = (rt == "MT" or rt == "MTd")

        self._cmake.configure()
        return self._cmake

    def package(self):
        cmake = self._configure_cmake()
        self.output.info(cmake.definitions)
        if self.logr_build_test_and_others:
            # If test are enabled build project and run tests.
            cmake.build()
            cmake.test(output_on_failure=True)

        cmake.install()

    def package_id(self):
        self.info.header_only()

    def package_info(self):
        self.info.header_only()
